Automattic/wp-calypso
# User: Refactor logout utility to use Redux

#### Changes proposed in this Pull Request

The last remaining legacy `lib/user/utils` function is `logout()`. This PR makes it a bit more independent by allowing it to accept an optional user prop and updates it to use Redux. It updates all locations to use that prop and removes the old function. Finally, we add a couple of tests to ensure the new functionality that uses `getLogoutUrl` without a user object still works.

Part of #24004 where we aim to reduxify `lib/user`.

#### Testing instructions

* Spin up Jetpack Cloud locally: `yarn run start-jetpack-cloud`
* Open `http://jetpack.cloud.localhost:3000/`
* Finish oauth if you haven't yet, pick a site.
* Click the profile button at the top right of your screen. 
* Click the link to logout.
* Verify you're redirected to Jetpack.com, and when you open `http://jetpack.cloud.localhost:3000/` again, it asks you to reauth.
* Start Calypso as usual with `yarn start` and continue with the next testing steps.
* Verify you are logged into WP.com.
* Go to http://calypso.localhost:3000/jetpack/connect/authorize?client_id=123123&site=example.com&nonce=nonce&secret=1&_wp_nonce=something&blogname=something&home_url=something&redirect_uri=something&scope=something&secret=something&site=something&site_url=something&state=something&woodna_service_name=login&debug
* Input your email address.
* On the next step, click "Create a new account"
* Verify it works like before and without errors in the console.
* Go through https://github.com/Automattic/wp-calypso/pull/53661#issuecomment-860828855 for setting up to test the desktop app dev version
* Verify logging out of the app still works the same way.
* Go to `/me/account` and close your account (best testable with a new fresh account without any upgrades) and verify it still works as it did before.
* Make sure you're logged in with a user that has 2FA enabled.
* Delete the `twostep_auth` cookie for `public-api.wordpress.com`.
* Go to `/me` and you'll see the reauthentication popup.
* Click the logout URL and verify it logs you out successfully, taking you to the WP.com landing page.
* Log in again and go to `/me`.
* Verify the log-out button in the sidebar works as it did before.
* Go through the invite flow and verify it still works well for both logged out and logged in users. Try also with a user that doesn't have a WP.com account yet.
